src/kademlia/dht_storage.cpp

```diff
@@ -0,0 +1,568 @@
+				}
+				m_map.erase(candidate);
+				m_counters.inc_stats_counter(counters::dht_torrents, -1);
+			}
+			m_counters.inc_stats_counter(counters::dht_torrents);
+			v = &m_map[info_hash];
+		}
+		else
+		{
+			v = &ti->second;
+		}
+
+		// the peer announces a torrent name, and we don't have a name
+		// for this torrent. Store it.
+		if (!name.empty() && v->name.empty())
+		{
+			std::string tname = name;
+			if (tname.size() > 50) tname.resize(50);
+			v->name = tname;
+		}
+
+		peer_entry peer;
+		peer.addr = endp;
+		peer.added = aux::time_now();
+		peer.seed = seed;
+		std::set<peer_entry>::iterator i = v->peers.find(peer);
+		if (i != v->peers.end()) v->peers.erase(i++);
+		v->peers.insert(i, peer);
+	}
+
+	bool dht_default_storage::get_immutable_item(sha1_hash const& target
+		, entry& item) const
+	{
+		dht_immutable_table_t::const_iterator i = m_immutable_table.find(target);
+		if (i == m_immutable_table.end()) return false;
+
+		item["v"] = bdecode(i->second.value, i->second.value + i->second.size);
+		return true;
+	}
+
+	void dht_default_storage::put_immutable_item(sha1_hash const& target
+		, char const* buf, int size
+		, address const& addr)
+	{
+		dht_immutable_table_t::iterator i = m_immutable_table.find(target);
+		if (i == m_immutable_table.end())
+		{
+			// make sure we don't add too many items
+			if (int(m_immutable_table.size()) >= m_settings.max_dht_items)
+			{
+				// delete the least important one (i.e. the one
+				// the fewest peers are announcing, and farthest
+				// from our node ID)
+				dht_immutable_table_t::iterator j = std::min_element(m_immutable_table.begin()
+					, m_immutable_table.end()
+					, immutable_item_comparator(m_id));
+
+				TORRENT_ASSERT(j != m_immutable_table.end());
+				free(j->second.value);
+				m_immutable_table.erase(j);
+				m_counters.inc_stats_counter(counters::dht_immutable_data, -1);
+			}
+			dht_immutable_item to_add;
+			to_add.value = static_cast<char*>(malloc(size));
+			to_add.size = size;
+			memcpy(to_add.value, buf, size);
+
+			boost::tie(i, boost::tuples::ignore) = m_immutable_table.insert(
+				std::make_pair(target, to_add));
+			m_counters.inc_stats_counter(counters::dht_immutable_data);
+		}
+
+//		fprintf(stderr, "added immutable item (%d)\n", int(m_immutable_table.size()));
+
+		touch_item(&i->second, addr);
+	}
+
+	bool dht_default_storage::get_mutable_item_seq(sha1_hash const& target
+		, boost::int64_t& seq) const
+	{
+		dht_mutable_table_t::const_iterator i = m_mutable_table.find(target);
+		if (i == m_mutable_table.end()) return false;
+
+		seq = i->second.seq;
+		return true;
+	}
+
+	bool dht_default_storage::get_mutable_item(sha1_hash const& target
+		, boost::int64_t seq, bool force_fill
+		, entry& item) const
+	{
+		dht_mutable_table_t::const_iterator i = m_mutable_table.find(target);
+		if (i == m_mutable_table.end()) return false;
+
+		dht_mutable_item const& f = i->second;
+		item["seq"] = f.seq;
+		if (force_fill || (0 <= seq && seq < f.seq))
```

Comment:
this expected behavior should probably be described in more detail in the documentation (in dht_storage_interface)